File: Hubs/Types/MarketReference.cs
using CryptoBot.Exchanges;
using MessagePack;
// using MessagePack;

namespace CryptoBotUI.Hubs.Types
{
    [MessagePackObject]
    public class MarketReference
    {
        [Key("exchangeName")]
        public string ExchangeName;

        [Key("symbol")]
        public string Symbol;

        [Key("key")]
        public string Key;

        public MarketReference() { }

        public MarketReference(Market market)
        {
            ExchangeName = market.Exchange.Name;
            Symbol = market.Pair.ToGenericSymbol();
            Key = $"{ExchangeName}::{Symbol}";
        }
    }
}

File: ClientApp/src/crypto-charts/lib/MarketReference.ts
export class MarketReference {
    public exchangeName: string;
    public symbol: string;
    public key: string;

    constructor(exchangeName: string, symbol: string) {
        this.exchangeName = exchangeName;
        this.symbol = symbol;
        this.key = `${this.exchangeName}::${this.symbol}`;
    }

    public get baseCurrency(): string {
        return this.symbol.split("/")[0];
    }
}